File: backend/app/api/v1/endpoints/exercise_library.py
from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session, joinedload
from sqlalchemy import func, or_, and_, text
from app.models_db import (
    ExerciseLibrary, ExerciseCategory, MuscleGroup, EquipmentType,
    ExerciseMuscleGroup, ExerciseEquipment, ExerciseVariation
)
from app.schemas.exercise_library import (
    ExerciseLibraryResponse, ExerciseLibraryCreate, ExerciseSearchParams,
    ExerciseSearchResponse, ExerciseCategoryResponse, MuscleGroupResponse,
    EquipmentTypeResponse, ExerciseQuickSelectResponse, ExerciseQuickSelect
)
from app.database.base import get_db
from typing import List, Optional
import uuid


router = APIRouter()


# --- Exercise Library Management ---

@router.get("/library", response_model=ExerciseSearchResponse)
def search_exercises(
    params: ExerciseSearchParams = Depends(),
    db: Session = Depends(get_db)
):
    """
    Advanced exercise search with filtering, pagination, and sorting.
    Supports full-text search, muscle group filtering, equipment filtering, etc.
    """
    
    # Base query with eager loading for relationships
    query = db.query(ExerciseLibrary).options(
        joinedload(ExerciseLibrary.category),
        joinedload(ExerciseLibrary.muscle_groups).joinedload(ExerciseMuscleGroup.muscle_group),
        joinedload(ExerciseLibrary.equipment).joinedload(ExerciseEquipment.equipment)
    ).filter(ExerciseLibrary.is_active == True)
    
    # Text search across name, description, and instructions
    if params.q:
        search_filter = or_(
            ExerciseLibrary.name.ilike(f"%{params.q}%"),
            ExerciseLibrary.description.ilike(f"%{params.q}%"),
            ExerciseLibrary.instructions.ilike(f"%{params.q}%")
        )
        query = query.filter(search_filter)
    
    # Category filtering
    if params.categories:
        category_ids = db.query(ExerciseCategory.id).filter(
            ExerciseCategory.name.in_(params.categories)
        ).subquery()
        query = query.filter(ExerciseLibrary.category_id.in_(category_ids))
    
    # Muscle group filtering
    if params.muscle_groups:
        muscle_group_ids = db.query(MuscleGroup.id).filter(
            MuscleGroup.name.in_(params.muscle_groups)
        ).subquery()
        exercise_ids_with_muscle_groups = db.query(ExerciseMuscleGroup.exercise_id).filter(
            ExerciseMuscleGroup.muscle_group_id.in_(muscle_group_ids)
        ).subquery()
        query = query.filter(ExerciseLibrary.id.in_(exercise_ids_with_muscle_groups))
    
    # Equipment filtering
    if params.equipment:
        equipment_ids = db.query(EquipmentType.id).filter(
            EquipmentType.name.in_(params.equipment)
        ).subquery()
        exercise_ids_with_equipment = db.query(ExerciseEquipment.exercise_id).filter(
            ExerciseEquipment.equipment_id.in_(equipment_ids)
        ).subquery()
        query = query.filter(ExerciseLibrary.id.in_(exercise_ids_with_equipment))
    
    # Difficulty filtering
    if params.difficulty:
        query = query.filter(ExerciseLibrary.difficulty_level.in_(params.difficulty))
    
    # Compound/unilateral filtering
    if params.compound is not None:
        query = query.filter(ExerciseLibrary.is_compound == params.compound)
    
    if params.unilateral is not None:
        query = query.filter(ExerciseLibrary.is_unilateral == params.unilateral)
    
    # Popularity filtering
    if params.min_popularity is not None:
        query = query.filter(ExerciseLibrary.popularity_score >= params.min_popularity)
    
    # Get total count before pagination
    total_count = query.count()
    
    # Sorting
    if params.sort_by == "name":
        order_column = ExerciseLibrary.name
    elif params.sort_by == "difficulty":
        order_column = ExerciseLibrary.difficulty_level
    elif params.sort_by == "popularity":
        order_column = ExerciseLibrary.popularity_score
    elif params.sort_by == "created_at":
        order_column = ExerciseLibrary.created_at
    else:
        order_column = ExerciseLibrary.popularity_score
    
    if params.sort_order == "asc":
        query = query.order_by(order_column.asc())
    else:
        query = query.order_by(order_column.desc())
    
    # Pagination
    offset = (params.page - 1) * params.page_size
    exercises = query.offset(offset).limit(params.page_size).all()
    
    total_pages = (total_count + params.page_size - 1) // params.page_size
    
    return ExerciseSearchResponse(
        exercises=exercises,
        total_count=total_count,
        page=params.page,
        page_size=params.page_size,
        total_pages=total_pages
    )


@router.get("/library/{exercise_id}", response_model=ExerciseLibraryResponse)
def get_exercise_details(
    exercise_id: uuid.UUID,
    db: Session = Depends(get_db)
):
    """Get detailed information about a specific exercise including variations."""
    
    exercise = db.query(ExerciseLibrary).options(
        joinedload(ExerciseLibrary.category),
        joinedload(ExerciseLibrary.muscle_groups).joinedload(ExerciseMuscleGroup.muscle_group),
        joinedload(ExerciseLibrary.equipment).joinedload(ExerciseEquipment.equipment),
        joinedload(ExerciseLibrary.variations_as_base).joinedload(ExerciseVariation.variation_exercise)
    ).filter(
        ExerciseLibrary.id == exercise_id,
        ExerciseLibrary.is_active == True
    ).first()
    
    if not exercise:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Exercise not found"
        )
    
    return exercise


@router.get("/quick-select", response_model=ExerciseQuickSelectResponse)
def get_exercises_for_selection(
    search: Optional[str] = Query(None, description="Search term"),
    category: Optional[str] = Query(None, description="Filter by category"),
    muscle_group: Optional[str] = Query(None, description="Filter by muscle group"),
    equipment: Optional[str] = Query(None, description="Filter by equipment"),
    limit: int = Query(50, ge=1, le=100, description="Number of exercises to return"),
    db: Session = Depends(get_db)
):
    """
    Get exercises in a simplified format for workout creation UI.
    Optimized for mobile app exercise selection with minimal data transfer.
    """
    
    # Base query for exercises
    query = db.query(ExerciseLibrary).options(
        joinedload(ExerciseLibrary.category),
        joinedload(ExerciseLibrary.muscle_groups).joinedload(ExerciseMuscleGroup.muscle_group),
        joinedload(ExerciseLibrary.equipment).joinedload(ExerciseEquipment.equipment)
    ).filter(ExerciseLibrary.is_active == True)
    
    # Apply filters
    if search:
        query = query.filter(ExerciseLibrary.name.ilike(f"%{search}%"))
    
    if category:
        query = query.join(ExerciseCategory).filter(ExerciseCategory.name.ilike(f"%{category}%"))
    
    if muscle_group:
        query = query.join(ExerciseMuscleGroup).join(MuscleGroup).filter(
            MuscleGroup.name.ilike(f"%{muscle_group}%")
        )
    
    if equipment:
        query = query.join(ExerciseEquipment).join(EquipmentType).filter(
            EquipmentType.name.ilike(f"%{equipment}%")
        )
    
    # Order by popularity and limit
    exercises = query.order_by(ExerciseLibrary.popularity_score.desc()).limit(limit).all()
    
    # Transform to quick select format
    quick_exercises = []
    for exercise in exercises:
        # Get primary muscle groups
        primary_muscles = [
            emg.muscle_group.name for emg in exercise.muscle_groups 
            if emg.is_primary
        ]
        
        # Get required equipment
        required_equipment = [
            ee.equipment.name for ee in exercise.equipment 
            if ee.is_required
        ]
        
        quick_exercises.append(ExerciseQuickSelect(
            id=exercise.id,
            name=exercise.name,
            category=exercise.category.name if exercise.category else None,
            difficulty_level=exercise.difficulty_level,
            primary_muscle_groups=primary_muscles,
            equipment_required=required_equipment,
            is_compound=exercise.is_compound
        ))
    
    # Get reference data for filters
    categories = db.query(ExerciseCategory).order_by(ExerciseCategory.name).all()
    muscle_groups = db.query(MuscleGroup).filter(
        MuscleGroup.group_type == 'primary'
    ).order_by(MuscleGroup.name).all()
    equipment_types = db.query(EquipmentType).order_by(EquipmentType.name).all()
    
    return ExerciseQuickSelectResponse(
        exercises=quick_exercises,
        categories=categories,
        muscle_groups=muscle_groups,
        equipment_types=equipment_types
    )


# --- Reference Data Endpoints ---

@router.get("/categories", response_model=List[ExerciseCategoryResponse])
def get_exercise_categories(db: Session = Depends(get_db)):
    """Get all exercise categories for filtering UI."""
    return db.query(ExerciseCategory).order_by(ExerciseCategory.name).all()


@router.get("/muscle-groups", response_model=List[MuscleGroupResponse])
def get_muscle_groups(
    group_type: Optional[str] = Query(None, description="Filter by group type: primary, secondary, stabilizer"),
    db: Session = Depends(get_db)
):
    """Get muscle groups, optionally filtered by type."""
    query = db.query(MuscleGroup).options(joinedload(MuscleGroup.children))
    
    if group_type:
        query = query.filter(MuscleGroup.group_type == group_type)
    
    return query.order_by(MuscleGroup.name).all()


@router.get("/equipment", response_model=List[EquipmentTypeResponse])
def get_equipment_types(
    category: Optional[str] = Query(None, description="Filter by equipment category"),
    db: Session = Depends(get_db)
):
    """Get equipment types, optionally filtered by category."""
    query = db.query(EquipmentType)
    
    if category:
        query = query.filter(EquipmentType.category == category)
    
    return query.order_by(EquipmentType.name).all()


# --- Exercise Analytics ---

@router.get("/popular", response_model=List[ExerciseLibraryResponse])
def get_popular_exercises(
    limit: int = Query(10, ge=1, le=50, description="Number of exercises to return"),
    category: Optional[str] = Query(None, description="Filter by category"),
    db: Session = Depends(get_db)
):
    """Get most popular exercises, optionally filtered by category."""
    
    query = db.query(ExerciseLibrary).options(
        joinedload(ExerciseLibrary.category),
        joinedload(ExerciseLibrary.muscle_groups).joinedload(ExerciseMuscleGroup.muscle_group),
        joinedload(ExerciseLibrary.equipment).joinedload(ExerciseEquipment.equipment)
    ).filter(ExerciseLibrary.is_active == True)
    
    if category:
        query = query.join(ExerciseCategory).filter(ExerciseCategory.name == category)
    
    return query.order_by(ExerciseLibrary.popularity_score.desc()).limit(limit).all()


@router.get("/similar/{exercise_id}", response_model=List[ExerciseLibraryResponse])
def get_similar_exercises(
    exercise_id: uuid.UUID,
    limit: int = Query(5, ge=1, le=20, description="Number of similar exercises to return"),
    db: Session = Depends(get_db)
):
    """Get exercises similar to the specified exercise based on muscle groups and equipment."""
    
    # Get the base exercise
    base_exercise = db.query(ExerciseLibrary).filter(
        ExerciseLibrary.id == exercise_id,
        ExerciseLibrary.is_active == True
    ).first()
    
    if not base_exercise:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Exercise not found"
        )
    
    # Get muscle groups and equipment for the base exercise
    base_muscle_groups = db.query(ExerciseMuscleGroup.muscle_group_id).filter(
        ExerciseMuscleGroup.exercise_id == exercise_id
    ).subquery()
    
    base_equipment = db.query(ExerciseEquipment.equipment_id).filter(
        ExerciseEquipment.exercise_id == exercise_id
    ).subquery()
    
    # Find similar exercises based on shared muscle groups or equipment
    similar_query = db.query(ExerciseLibrary).options(
        joinedload(ExerciseLibrary.category),
        joinedload(ExerciseLibrary.muscle_groups).joinedload(ExerciseMuscleGroup.muscle_group),
        joinedload(ExerciseLibrary.equipment).joinedload(ExerciseEquipment.equipment)
    ).filter(
        ExerciseLibrary.id != exercise_id,
        ExerciseLibrary.is_active == True,
        or_(
            ExerciseLibrary.id.in_(
                db.query(ExerciseMuscleGroup.exercise_id).filter(
                    ExerciseMuscleGroup.muscle_group_id.in_(base_muscle_groups)
                )
            ),
            ExerciseLibrary.id.in_(
                db.query(ExerciseEquipment.exercise_id).filter(
                    ExerciseEquipment.equipment_id.in_(base_equipment)
                )
            )
        )
    ).order_by(ExerciseLibrary.popularity_score.desc()).limit(limit)
    
    return similar_query.all()


# --- Exercise Variations ---

@router.get("/variations/{exercise_id}", response_model=List[ExerciseLibraryResponse])
def get_exercise_variations(
    exercise_id: uuid.UUID,
    variation_type: Optional[str] = Query(None, description="Filter by variation type: easier, harder, alternative, progression"),
    db: Session = Depends(get_db)
):
    """Get exercise variations for a specific exercise."""
    
    query = db.query(ExerciseLibrary).join(
        ExerciseVariation, ExerciseLibrary.id == ExerciseVariation.variation_exercise_id
    ).options(
        joinedload(ExerciseLibrary.category),
        joinedload(ExerciseLibrary.muscle_groups).joinedload(ExerciseMuscleGroup.muscle_group),
        joinedload(ExerciseLibrary.equipment).joinedload(ExerciseEquipment.equipment)
    ).filter(
        ExerciseVariation.base_exercise_id == exercise_id,
        ExerciseLibrary.is_active == True
    )
    
    if variation_type:
        query = query.filter(ExerciseVariation.variation_type == variation_type)
    
    return query.all()

File: fitnation/lib/models/ExerciseLibrary.dart
// Enhanced Exercise Library Models for Fitnation App

class ExerciseCategory {
  final String id;
  final String name;
  final String? description;
  final String? colorCode;
  final String? iconName;
  final DateTime createdAt;
  final DateTime updatedAt;

  ExerciseCategory({
    required this.id,
    required this.name,
    this.description,
    this.colorCode,
    this.iconName,
    required this.createdAt,
    required this.updatedAt,
  });

  factory ExerciseCategory.fromJson(Map<String, dynamic> json) {
    return ExerciseCategory(
      id: json['id'] as String,
      name: json['name'] as String,
      description: json['description'] as String?,
      colorCode: json['color_code'] as String?,
      iconName: json['icon_name'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      if (description != null) 'description': description,
      if (colorCode != null) 'color_code': colorCode,
      if (iconName != null) 'icon_name': iconName,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
    };
  }
}

class MuscleGroup {
  final String id;
  final String name;
  final String groupType; // primary, secondary, stabilizer
  final String? parentId;
  final String? description;
  final DateTime createdAt;
  final DateTime updatedAt;
  final List<MuscleGroup> children;

  MuscleGroup({
    required this.id,
    required this.name,
    required this.groupType,
    this.parentId,
    this.description,
    required this.createdAt,
    required this.updatedAt,
    this.children = const [],
  });

  factory MuscleGroup.fromJson(Map<String, dynamic> json) {
    return MuscleGroup(
      id: json['id'] as String,
      name: json['name'] as String,
      groupType: json['group_type'] as String,
      parentId: json['parent_id'] as String?,
      description: json['description'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
      children:
          (json['children'] as List<dynamic>?)
              ?.map(
                (child) => MuscleGroup.fromJson(child as Map<String, dynamic>),
              )
              .toList() ??
          [],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'group_type': groupType,
      if (parentId != null) 'parent_id': parentId,
      if (description != null) 'description': description,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
      'children': children.map((child) => child.toJson()).toList(),
    };
  }
}

class EquipmentType {
  final String id;
  final String name;
  final String? category;
  final String? description;
  final DateTime createdAt;
  final DateTime updatedAt;

  EquipmentType({
    required this.id,
    required this.name,
    this.category,
    this.description,
    required this.createdAt,
    required this.updatedAt,
  });

  factory EquipmentType.fromJson(Map<String, dynamic> json) {
    return EquipmentType(
      id: json['id'] as String,
      name: json['name'] as String,
      category: json['category'] as String?,
      description: json['description'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      if (category != null) 'category': category,
      if (description != null) 'description': description,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
    };
  }
}

class ExerciseMuscleGroup {
  final String id;
  final String muscleGroupId;
  final bool isPrimary;
  final int? activationLevel; // 1-5 scale
  final MuscleGroup muscleGroup;

  ExerciseMuscleGroup({
    required this.id,
    required this.muscleGroupId,
    required this.isPrimary,
    this.activationLevel,
    required this.muscleGroup,
  });

  factory ExerciseMuscleGroup.fromJson(Map<String, dynamic> json) {
    return ExerciseMuscleGroup(
      id: json['id'] as String,
      muscleGroupId: json['muscle_group_id'] as String,
      isPrimary: json['is_primary'] as bool,
      activationLevel: json['activation_level'] as int?,
      muscleGroup: MuscleGroup.fromJson(
        json['muscle_group'] as Map<String, dynamic>,
      ),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'muscle_group_id': muscleGroupId,
      'is_primary': isPrimary,
      if (activationLevel != null) 'activation_level': activationLevel,
      'muscle_group': muscleGroup.toJson(),
    };
  }
}

class ExerciseEquipment {
  final String id;
  final String equipmentId;
  final bool isRequired;
  final bool isAlternative;
  final EquipmentType equipment;

  ExerciseEquipment({
    required this.id,
    required this.equipmentId,
    required this.isRequired,
    required this.isAlternative,
    required this.equipment,
  });

  factory ExerciseEquipment.fromJson(Map<String, dynamic> json) {
    return ExerciseEquipment(
      id: json['id'] as String,
      equipmentId: json['equipment_id'] as String,
      isRequired: json['is_required'] as bool,
      isAlternative: json['is_alternative'] as bool,
      equipment: EquipmentType.fromJson(
        json['equipment'] as Map<String, dynamic>,
      ),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'equipment_id': equipmentId,
      'is_required': isRequired,
      'is_alternative': isAlternative,
      'equipment': equipment.toJson(),
    };
  }
}

class ExerciseVariation {
  final String id;
  final String variationExerciseId;
  final String? variationType; // easier, harder, alternative, progression
  final String? notes;
  final ExerciseLibrary variationExercise;

  ExerciseVariation({
    required this.id,
    required this.variationExerciseId,
    this.variationType,
    this.notes,
    required this.variationExercise,
  });

  factory ExerciseVariation.fromJson(Map<String, dynamic> json) {
    return ExerciseVariation(
      id: json['id'] as String,
      variationExerciseId: json['variation_exercise_id'] as String,
      variationType: json['variation_type'] as String?,
      notes: json['notes'] as String?,
      variationExercise: ExerciseLibrary.fromJson(
        json['variation_exercise'] as Map<String, dynamic>,
      ),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'variation_exercise_id': variationExerciseId,
      if (variationType != null) 'variation_type': variationType,
      if (notes != null) 'notes': notes,
      'variation_exercise': variationExercise.toJson(),
    };
  }
}

class ExerciseLibrary {
  final String id;
  final String name;
  final String? description;
  final String? instructions;
  final String? formTips;
  final String? gifUrl;
  final String? videoUrl;
  final String? categoryId;
  final int? difficultyLevel; // 1-5 scale
  final double popularityScore; // 0.0-5.0
  final bool isCompound;
  final bool isUnilateral;
  final bool isActive;
  final DateTime createdAt;
  final DateTime updatedAt;

  // Relationships
  final ExerciseCategory? category;
  final List<ExerciseMuscleGroup> muscleGroups;
  final List<ExerciseEquipment> equipment;
  final List<ExerciseVariation> variationsAsBase;

  ExerciseLibrary({
    required this.id,
    required this.name,
    this.description,
    this.instructions,
    this.formTips,
    this.gifUrl,
    this.videoUrl,
    this.categoryId,
    this.difficultyLevel,
    this.popularityScore = 0.0,
    this.isCompound = false,
    this.isUnilateral = false,
    this.isActive = true,
    required this.createdAt,
    required this.updatedAt,
    this.category,
    this.muscleGroups = const [],
    this.equipment = const [],
    this.variationsAsBase = const [],
  });

  factory ExerciseLibrary.fromJson(Map<String, dynamic> json) {
    return ExerciseLibrary(
      id: json['id'] as String,
      name: json['name'] as String,
      description: json['description'] as String?,
      instructions: json['instructions'] as String?,
      formTips: json['form_tips'] as String?,
      gifUrl: json['gif_url'] as String?,
      videoUrl: json['video_url'] as String?,
      categoryId: json['category_id'] as String?,
      difficultyLevel: json['difficulty_level'] as int?,
      popularityScore: (json['popularity_score'] as num?)?.toDouble() ?? 0.0,
      isCompound: json['is_compound'] as bool? ?? false,
      isUnilateral: json['is_unilateral'] as bool? ?? false,
      isActive: json['is_active'] as bool? ?? true,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
      category:
          json['category'] != null
              ? ExerciseCategory.fromJson(
                json['category'] as Map<String, dynamic>,
              )
              : null,
      muscleGroups:
          (json['muscle_groups'] as List<dynamic>?)
              ?.map(
                (mg) =>
                    ExerciseMuscleGroup.fromJson(mg as Map<String, dynamic>),
              )
              .toList() ??
          [],
      equipment:
          (json['equipment'] as List<dynamic>?)
              ?.map(
                (eq) => ExerciseEquipment.fromJson(eq as Map<String, dynamic>),
              )
              .toList() ??
          [],
      variationsAsBase:
          (json['variations_as_base'] as List<dynamic>?)
              ?.map(
                (variation) => ExerciseVariation.fromJson(
                  variation as Map<String, dynamic>,
                ),
              )
              .toList() ??
          [],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      if (description != null) 'description': description,
      if (instructions != null) 'instructions': instructions,
      if (formTips != null) 'form_tips': formTips,
      if (gifUrl != null) 'gif_url': gifUrl,
      if (videoUrl != null) 'video_url': videoUrl,
      if (categoryId != null) 'category_id': categoryId,
      if (difficultyLevel != null) 'difficulty_level': difficultyLevel,
      'popularity_score': popularityScore,
      'is_compound': isCompound,
      'is_unilateral': isUnilateral,
      'is_active': isActive,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
      if (category != null) 'category': category!.toJson(),
      'muscle_groups': muscleGroups.map((mg) => mg.toJson()).toList(),
      'equipment': equipment.map((eq) => eq.toJson()).toList(),
      'variations_as_base':
          variationsAsBase.map((variation) => variation.toJson()).toList(),
    };
  }

  // Helper methods for UI
  List<String> get primaryMuscleGroupNames {
    return muscleGroups
        .where((mg) => mg.isPrimary)
        .map((mg) => mg.muscleGroup.name)
        .toList();
  }

  List<String> get secondaryMuscleGroupNames {
    return muscleGroups
        .where((mg) => !mg.isPrimary)
        .map((mg) => mg.muscleGroup.name)
        .toList();
  }

  List<String> get requiredEquipmentNames {
    return equipment
        .where((eq) => eq.isRequired)
        .map((eq) => eq.equipment.name)
        .toList();
  }

  List<String> get alternativeEquipmentNames {
    return equipment
        .where((eq) => eq.isAlternative)
        .map((eq) => eq.equipment.name)
        .toList();
  }

  String get difficultyText {
    switch (difficultyLevel) {
      case 1:
        return 'Beginner';
      case 2:
        return 'Easy';
      case 3:
        return 'Intermediate';
      case 4:
        return 'Hard';
      case 5:
        return 'Expert';
      default:
        return 'Unknown';
    }
  }

  String get categoryName => category?.name ?? 'Uncategorized';

  String get displayName => name;

  String get primaryEquipment =>
      requiredEquipmentNames.isNotEmpty
          ? requiredEquipmentNames.first
          : 'Body Weight';
}

// --- Search and Filter Models ---

class ExerciseSearchFilters {
  final String? searchQuery;
  final List<String> categoryIds;
  final List<String> muscleGroupIds;
  final List<String> equipmentIds;
  final List<int> difficultyLevels;
  final bool? isCompound;
  final bool? isUnilateral;
  final double? minPopularity;
  final bool equipmentRequiredOnly;

  ExerciseSearchFilters({
    this.searchQuery,
    this.categoryIds = const [],
    this.muscleGroupIds = const [],
    this.equipmentIds = const [],
    this.difficultyLevels = const [],
    this.isCompound,
    this.isUnilateral,
    this.minPopularity,
    this.equipmentRequiredOnly = false,
  });

  Map<String, dynamic> toQueryParams() {
    final params = <String, dynamic>{};

    if (searchQuery != null && searchQuery!.isNotEmpty) {
      params['q'] = searchQuery;
    }

    if (categoryIds.isNotEmpty) {
      params['categories'] = categoryIds;
    }

    if (muscleGroupIds.isNotEmpty) {
      params['muscle_groups'] = muscleGroupIds;
    }

    if (equipmentIds.isNotEmpty) {
      params['equipment'] = equipmentIds;
    }

    if (difficultyLevels.isNotEmpty) {
      params['difficulty'] = difficultyLevels;
    }

    if (isCompound != null) {
      params['compound'] = isCompound;
    }

    if (isUnilateral != null) {
      params['unilateral'] = isUnilateral;
    }

    if (minPopularity != null) {
      params['min_popularity'] = minPopularity;
    }

    return params;
  }

  ExerciseSearchFilters copyWith({
    String? searchQuery,
    List<String>? categoryIds,
    List<String>? muscleGroupIds,
    List<String>? equipmentIds,
    List<int>? difficultyLevels,
    bool? isCompound,
    bool? isUnilateral,
    double? minPopularity,
    bool? equipmentRequiredOnly,
  }) {
    return ExerciseSearchFilters(
      searchQuery: searchQuery ?? this.searchQuery,
      categoryIds: categoryIds ?? this.categoryIds,
      muscleGroupIds: muscleGroupIds ?? this.muscleGroupIds,
      equipmentIds: equipmentIds ?? this.equipmentIds,
      difficultyLevels: difficultyLevels ?? this.difficultyLevels,
      isCompound: isCompound ?? this.isCompound,
      isUnilateral: isUnilateral ?? this.isUnilateral,
      minPopularity: minPopularity ?? this.minPopularity,
      equipmentRequiredOnly:
          equipmentRequiredOnly ?? this.equipmentRequiredOnly,
    );
  }
}

class ExerciseSearchResponse {
  final List<ExerciseLibrary> exercises;
  final int totalCount;
  final int page;
  final int pageSize;
  final int totalPages;

  ExerciseSearchResponse({
    required this.exercises,
    required this.totalCount,
    required this.page,
    required this.pageSize,
    required this.totalPages,
  });

  factory ExerciseSearchResponse.fromJson(Map<String, dynamic> json) {
    return ExerciseSearchResponse(
      exercises:
          (json['exercises'] as List<dynamic>)
              .map((ex) => ExerciseLibrary.fromJson(ex as Map<String, dynamic>))
              .toList(),
      totalCount: json['total_count'] as int,
      page: json['page'] as int,
      pageSize: json['page_size'] as int,
      totalPages: json['total_pages'] as int,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'exercises': exercises.map((ex) => ex.toJson()).toList(),
      'total_count': totalCount,
      'page': page,
      'page_size': pageSize,
      'total_pages': totalPages,
    };
  }
}

// --- Quick Selection Models for Workout Creation ---

class ExerciseQuickSelect {
  final String id;
  final String name;
  final String? category;
  final int? difficultyLevel;
  final List<String> primaryMuscleGroups;
  final List<String> equipmentRequired;
  final bool isCompound;

  ExerciseQuickSelect({
    required this.id,
    required this.name,
    this.category,
    this.difficultyLevel,
    this.primaryMuscleGroups = const [],
    this.equipmentRequired = const [],
    this.isCompound = false,
  });

  factory ExerciseQuickSelect.fromJson(Map<String, dynamic> json) {
    return ExerciseQuickSelect(
      id: json['id'] as String,
      name: json['name'] as String,
      category: json['category'] as String?,
      difficultyLevel: json['difficulty_level'] as int?,
      primaryMuscleGroups:
          (json['primary_muscle_groups'] as List<dynamic>?)
              ?.map((mg) => mg as String)
              .toList() ??
          [],
      equipmentRequired:
          (json['equipment_required'] as List<dynamic>?)
              ?.map((eq) => eq as String)
              .toList() ??
          [],
      isCompound: json['is_compound'] as bool? ?? false,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      if (category != null) 'category': category,
      if (difficultyLevel != null) 'difficulty_level': difficultyLevel,
      'primary_muscle_groups': primaryMuscleGroups,
      'equipment_required': equipmentRequired,
      'is_compound': isCompound,
    };
  }
}

class ExerciseQuickSelectResponse {
  final List<ExerciseQuickSelect> exercises;
  final List<ExerciseCategory> categories;
  final List<MuscleGroup> muscleGroups;
  final List<EquipmentType> equipmentTypes;

  ExerciseQuickSelectResponse({
    required this.exercises,
    required this.categories,
    required this.muscleGroups,
    required this.equipmentTypes,
  });

  factory ExerciseQuickSelectResponse.fromJson(Map<String, dynamic> json) {
    return ExerciseQuickSelectResponse(
      exercises:
          (json['exercises'] as List<dynamic>)
              .map(
                (ex) =>
                    ExerciseQuickSelect.fromJson(ex as Map<String, dynamic>),
              )
              .toList(),
      categories:
          (json['categories'] as List<dynamic>)
              .map(
                (cat) => ExerciseCategory.fromJson(cat as Map<String, dynamic>),
              )
              .toList(),
      muscleGroups:
          (json['muscle_groups'] as List<dynamic>)
              .map((mg) => MuscleGroup.fromJson(mg as Map<String, dynamic>))
              .toList(),
      equipmentTypes:
          (json['equipment_types'] as List<dynamic>)
              .map((et) => EquipmentType.fromJson(et as Map<String, dynamic>))
              .toList(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'exercises': exercises.map((ex) => ex.toJson()).toList(),
      'categories': categories.map((cat) => cat.toJson()).toList(),
      'muscle_groups': muscleGroups.map((mg) => mg.toJson()).toList(),
      'equipment_types': equipmentTypes.map((et) => et.toJson()).toList(),
    };
  }
}
